### PR TITLE
feat: Allow restriction of number of inputs per tx

### PR DESCRIPTION
**Pull request by Sword-Smith**
Allow the user to restrict the number of inputs per transaction to avoid having to deal with transactions that will likely never be mined, as they don't fit into a block.

The current default limit is set quite low, to 28 inputs per transaction due to the big size of transaction inputs. This limit is intended to rise significantly in the future, but for now it should serve to prevent composers and transaction initiators from making transactions that cannot be mined.

Both the wallet state and mempool state gets passed this value such that they, internally, can enforce the correct policies.

### CODE REVIEW

**Comment by Sword-Smith** [comment]
Marked as draft because this new parameter is completely untested. Needs tests!

File: src/config_models/cli_args.rs [hunk 1]
```diff
@@ -170,6 +170,21 @@ pub struct Args {
...
+    /// Restrict the number of inputs for transcations that this node deals
+    /// with.
+    ///
+    /// If this value is set, the node will not create transactions with more
+    /// inputs than this limit. It will also not accept such transactions from
+    /// peers into its mempool. And if the node is either upgrading transaction
+    /// proofs on the network or composing, it will never merge two transactions
+    /// that would exceed this limit. Does not affect block validity, i.e.
+    /// consensus. So a block with more than this number of inputs may be
+    /// considered value. Also does not affect which block proposals that a
```
Comment (aszepieniec) [inline src/config_models/cli_args.rs:182]:
```suggestion
    /// consensus: a block with more than this number of inputs can be
    /// valid. Also does not affect which block proposals that a
```

File: src/config_models/cli_args.rs [hunk 2]
```diff
@@ -170,6 +170,21 @@ pub struct Args {
...
+    /// guessing node will attempt to solve the PoW puzzle for. Set to "none" to
+    /// disable this limit.
```
Comment (aszepieniec) [inline src/config_models/cli_args.rs:184]: If the default is not none, it should say here what it is.

File: src/config_models/cli_args.rs [hunk 3]
```diff
@@ -170,6 +170,21 @@ pub struct Args {
...
+    #[clap(long, value_parser = parse_optional_usize, default_value = "28")]
```
Comment (aszepieniec) [inline src/config_models/cli_args.rs:185]: It is not obvious to me that the default string will parse correctly. Please add it to the test case below.

File: src/main_loop.rs [hunk 4]
```diff
@@ -807,14 +807,18 @@ impl MainLoopHandler {
...
-                global_state_mut
+                if let Err(err) = global_state_mut
                     .mempool_insert(
                         pt2m_transaction.transaction.to_owned(),
                         TransactionOrigin::Foreign,
                     )
-                    .await;
+                    .await
+                {
+                    warn!("{err}");
```
Comment (aszepieniec) [inline src/main_loop.rs:817]: Add context:
```suggestion
                    warn!("cannot add transaction into mempool: {err}");
```

File: src/main_loop.rs [hunk 5]
```diff
@@ -807,14 +807,18 @@ impl MainLoopHandler {
                 }
 
                 // Insert into mempool
```
Comment (aszepieniec) [inline src/main_loop.rs:809]:
```suggestion
                // Insert into mempool or return prematurely
```

File: src/models/state/mempool.rs [hunk 6]
```diff
@@ -226,31 +238,48 @@ impl Mempool {
...
-                own_tx = own_tx || candidate.origin.is_own();
-            }
-
-            let origin = match own_tx {
-                true => TransactionOrigin::Own,
-                false => TransactionOrigin::Foreign,
-            };
+        /// Loop from `i`th most valuable transaction in the mempool to the
+        /// least valuable in order to find a pair whose total number of inputs
+        /// does not exceed the maximum value.
```
Comment (aszepieniec) [inline src/models/state/mempool.rs:251]:
```suggestion
        /// Loop from `i`th densest transaction in the mempool to the
        /// least dense in order to find a pair whose total number of inputs
        /// does not exceed the maximum value.
```

File: src/models/state/mempool.rs [hunk 7]
```diff
@@ -226,31 +238,48 @@ impl Mempool {
...
+        fn skip_first_n_transaction(
```
Comment (aszepieniec) [inline src/models/state/mempool.rs:252]: Something is missing from this name to describe what the function does.

File: src/models/state/mempool.rs [hunk 8]
```diff
@@ -346,13 +375,18 @@ impl Mempool {
             }
         }
 
-        let mut events = vec![];
+        // Check if tx conforms to mempool policy.
```
Comment (aszepieniec) [inline src/models/state/mempool.rs:378]:
```suggestion
        // Check if tx conforms to mempool policy wrt # inputs
```

File: src/models/state/mempool/mempool_insertion_error.rs [hunk 9]
```diff
@@ -0,0 +1,6 @@
+/// The reasons why inserting a transaction into the mempool may fail.
+#[derive(Debug, Clone, Copy, thiserror::Error)]
+pub(crate) enum MempoolInsertionError {
+    #[error("Too many inputs: got {got}, but only {allowed} are allowed.")]
+    TooManyInputs { got: usize, allowed: usize },
+}
```
Comment (aszepieniec) [inline src/models/state/mempool/mempool_insertion_error.rs:6]: Too small and intricately tied to the mempool module to justify its own file.

**Review by aszepieniec — COMMENTED** [review]
Basic structure looks good. I have some superficial comments. Before approving I would like to double-check against the list on the whiteboard and verify that the right logic is used in all of those cases.

File: src/models/state/mempool/mempool_insertion_error.rs (hunk 9 above)
Comment (Sword-Smith) [reply src/models/state/mempool/mempool_insertion_error.rs:6]: It might not be obvious from the code base but I rather like small files :)

File: src/config_models/cli_args.rs (hunk 2 above)
Comment (Sword-Smith) [reply src/config_models/cli_args.rs:184]: `clap` fills that out automatically when you use e.g. `neptune-core --help`. See this example for how it works with `--max-num-proofs`:
```bash
      --max-num-proofs <MAX_NUM_PROOFS>
          Sets the maximum number of proofs in a `ProofCollection` that can be recursively combined into a `SingleProof` by this machine. I.e. how big STARK proofs this machine can produce
          
          [default: 16]

```

File: src/models/state/mempool/mempool_insertion_error.rs (hunk 9 above)
Comment (jan-ferdinand) [reply src/models/state/mempool/mempool_insertion_error.rs:6]: It's good to see more use of `thiserror` instead of `anyhow` in the library. I advocate for re-exporting all error types in a top-level `error.rs` to simplify discovery for downstream users.

File: src/models/state/mempool.rs [hunk 10]
```diff
@@ -152,6 +162,7 @@ impl Mempool {
     pub fn new(
         max_total_size: ByteSize,
         max_num_transactions: Option<usize>,
+        max_num_inputs_per_tx: Option<usize>,
```
Comment (jan-ferdinand) [inline src/models/state/mempool.rs:165]: This interface is starting to get unwieldy. How about a builder instead? If you don't want to write one yourself, take a look at [bon](https://crates.io/crates/bon).

File: src/models/state/mempool.rs [hunk 11]
```diff
@@ -226,31 +239,55 @@ impl Mempool {
...
+                    };
+                    return Some((ret.try_into().unwrap(), origin));
+                }
             }
+
+            None
         }
 
-        None
+        (0..self.len() - 1).find_map(|i| most_dense_pair_skip_first_n(self, i))
```
Comment (Sword-Smith) [inline src/models/state/mempool.rs:290]: I think the expression `self.len() - 1` can overflow. Probably not good.

Comment (Sword-Smith) [reply src/models/state/mempool.rs:290]: Just shows that these getters need to be tested on an empty mempool. That would have caught this problem by producing a never-ending test case.
